optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +127,134 @@ private void initBenchmarkDirectoryAndSubdirs() {
     }
 
     private void warmUp() {
-        if (plannerBenchmarkResult.getWarmUpTimeMillisSpentLimit() > 0L) {
-            logger.info("================================================================================");
-            logger.info("Warm up started");
-            logger.info("================================================================================");
-            long startingTimeMillis = System.currentTimeMillis();
-            long timeLeft = plannerBenchmarkResult.getWarmUpTimeMillisSpentLimit();
-            List<ProblemBenchmarkResult> unifiedProblemBenchmarkResultList = plannerBenchmarkResult.getUnifiedProblemBenchmarkResultList();
-            Iterator<ProblemBenchmarkResult> it = unifiedProblemBenchmarkResultList.iterator();
-            while (timeLeft > 0L) {
-                if (!it.hasNext()) {
-                    it = unifiedProblemBenchmarkResultList.iterator();
-                }
-                ProblemBenchmarkResult problemBenchmarkResult = it.next();
-                timeLeft = warmUp(problemBenchmarkResult, startingTimeMillis, plannerBenchmarkResult.getWarmUpTimeMillisSpentLimit(), timeLeft);
+        if (plannerBenchmarkResult.getWarmUpTimeMillisSpentLimit() <= 0L) {
+            return;
+        }
+        logger.info("================================================================================");
+        logger.info("Warm up started");
+        logger.info("================================================================================");
+        long timeLeftTotal = plannerBenchmarkResult.getWarmUpTimeMillisSpentLimit();
+        int parallelBenchmarkCount = plannerBenchmarkResult.getParallelBenchmarkCount();
+        int solverBenchmarkResultCount = plannerBenchmarkResult.getSolverBenchmarkResultList().size();
+        long cyclesCount = Math.round(Math.ceil(solverBenchmarkResultCount / (double) parallelBenchmarkCount));
+        long timeLeftPerCycle = Math.round(Math.floor((timeLeftTotal / (double) cyclesCount)));
+        ConcurrentMap<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner> futureMap
+                = new ConcurrentHashMap<Future<SingleBenchmarkRunner>, SingleBenchmarkRunner>(parallelBenchmarkCount);
+        Map<SolverBenchmarkResult, TerminationConfig> originalTerminationConfigMap
+                = new HashMap<SolverBenchmarkResult, TerminationConfig>(solverBenchmarkResultCount);
+        ConcurrentMap<SolverBenchmarkResult, Integer> singleBenchmarkResultIndexMap
+                = new ConcurrentHashMap<SolverBenchmarkResult, Integer>(solverBenchmarkResultCount);
+
+        warmUpStarted(originalTerminationConfigMap);
+        SolverBenchmarkResult[] solverBenchmarkResultCycleArray = new SolverBenchmarkResult[parallelBenchmarkCount];
```

Comment:
Noted the convention, but in this case, `solverBenchmarkResultCycle` makes more sense, as its contents are replaced at the beginning of each single cycle.